lib/aws/cognito_srp.rb

```diff
@@ -115,11 +119,17 @@ def authenticate
 
       challenge_response = process_challenge(init_auth_response.challenge_parameters)
 
-      auth_response = @aws_client.respond_to_auth_challenge(
+      @user_id_for_srp = challenge_response[:USERNAME]
```

Comment:
This should be the same as `@username` and therefore no need for a separate instance variable, no?

Reply:
`@user_id_for_srp` is not same as `@username`. 
`@user_id_for_srp` is same as user sub(ID) of `@username`.

But it is not clear code. So, I refactor it.
https://github.com/beezwax/aws-cognito-srp-ruby/pull/2/commits/5f429f85782a59092684aeab1b579d5bfa443d25

Reply:
@pilaf 
> Would you be able to add specs for it too?
I'll try to add spec later.

Reply:
> therefore no need for a separate instance variable, no?

`@user_id_for_srp` is needed for `refresh_tokens`.

Reply:
Thanks @suketa 

I still think we shouldn't need this separate instance variable, as it makes `#refresh_tokens` dependent on having called `#authenticate` first to populate `@user_id_for_srp`, and that shouldn't be a requirement. `#refresh_tokens` should be able to work in isolation.

But that said, I'm also not convinced that `USER_ID_FOR_SRP` wouldn't be the same value as our `@username` set in `#initialize`. From [Cognito's documentation](https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-authentication-flow.html):

> In the `AdminInitiateAuth` response `ChallengeParameters`, the `USER_ID_FOR_SRP` attribute, if present, contains the user's actual user name, not an alias (such as email address or phone number). In your call to `AdminRespondToAuthChallenge`, in the `ChallengeResponses`, you must pass this user name in the `USERNAME` parameter.

So if I understand that right, `USER_ID_FOR_SRP` will only differ from our original `@username` when an alias (email or phone number) was used. Since you're only using `@user_id_for_srp` for the `SECRET_HASH`, which should be the same for the `USER_SRP_AUTH` and `REFRESH_TOKEN` auth flows, I think we can probably replace it with `@username`.

For additional confirmation I checked how Warrant (the Python library this gem is based on) deals with this, and it seems they just use the originally given username:

From its `renew_access_token` function ([code](https://github.com/capless/warrant/blob/e81f9a6dd6e36be3b6fa1dbbd11334ec92124ace/warrant/__init__.py#L550-L569))

```python
    def renew_access_token(self):
        """
        Sets a new access token on the User using the refresh token.
        """
        auth_params = {'REFRESH_TOKEN': self.refresh_token}
        self._add_secret_hash(auth_params, 'SECRET_HASH')
        # ...
```

And in turn `_add_secret_hash` ([code](https://github.com/capless/warrant/blob/e81f9a6dd6e36be3b6fa1dbbd11334ec92124ace/warrant/__init__.py#L623-L631)):

```python
    def _add_secret_hash(self, parameters, key):
        """
        Helper function that computes SecretHash and adds it
        to a parameters dictionary at a specified key
        """
        if self.client_secret is not None:
            secret_hash = AWSSRP.get_secret_hash(self.username, self.client_id,
                                                 self.client_secret)
            parameters[key] = secret_hash
```

Reply:
@pilaf Thank you.

My user pool accepts sign-in by email only. And not accept username.
So, In my case, **USER_ID_FOR_SRP is  differ from our original @username** because an alias email was used for sign-in.

And I got the following error in `refresh_token` with secret_hash created from original `@username`.
SecretHash does not match for the client: XXXXXXX (Aws::CognitoIdentityProvider::Errors::NotAuthorizedException)
But, it works fine with `@user_id_for_srp`.
So I think we need `@user_id_for_srp`.

> For additional confirmation I checked how Warrant (the Python library this gem is based on) deals with this, and it seems they just use the originally given username:

I have not tested, but I think Warrant has same issue. https://github.com/capless/warrant/issues/74

Reply:
@suketa I see, you seem to be right, thanks for the explanation.

In that case I think the `#refresh_tokens` method should take an additional argument so it can be given the `user_id_for_srp` when it's not present, since the call to `#authenticate` and `#refresh_tokens` will not always happen in the same instance.

Something like this perhaps?

```ruby
def refresh_tokens(refresh_token, user_id_for_srp: @user_id_for_srp)
```

Reply:
@pilaf 
Yes, it seems to be better. 
And one more, how about adding `#user_id_for_srp` method which returns `@user_id_for_srp` ?

Because in my case, there is no way to know `@user_id_for_srp` from email.

Reply:
@suketa 

> And one more, how about adding `#user_id_for_srp` method which returns `@user_id_for_srp`?

Do you mean like a simple `attr_reader`, or is there a Cognito API endpoint to get the `USER_ID_FOR_SRP` from an alias without going through an auth flow?

Reply:
@pilaf 
I mean adding a simple `attr_reader`

Reply:
@suketa Oh, definitely, go ahead.

Reply:
@pilaf Thank you